### DISCUSSION TITLE
Lower required PHP version but with a warning - thoughts?

### DISCUSSION THREAD

**Discussion by jonyo** [geodesicsolutions-community/geocore-community#66] [Ideas]
As mentioned in one of the issues, the only reason I set the min version to 7.4 was that earlier versions are no longer "officially" supported by PHP even for security fixes.

For context, the official PHP releases are supported for 2 years for active development, then 1 more year of "security fixes only", so the typical major version of PHP has a 3 year life cycle.  See https://www.php.net/supported-versions.php

But I remembered something, for one thing, some external projects (like Ubuntu I beleive) actually do maintain older versions if they are included in a LTS version of ubuntu.  So even if it says 7.2 for instance, that is not the same 7.2 version you'd see for download officially on php.net, it actually has all the more recent security patches on top.

Also when you raise the required version, you narrow the number of people that can use it some.  So while the community is still picking up steam it may be better to lower the threshold to hopefully attract more folks.

Soo with that in mind... What do you think of having a "soft limit" of PHP 7.4 - if you have a lower version, it still lets you update, but shows a warning that we recommend updating to the latest version of 7.4?

**Comment by vicos59** [comment]
I agree with not being too restrictive. There are also special Linux Web Server packages that offer hardened versions of old PHP back to 5.4.

I just reviewed Xenforo's requirements and they do this:

Min: 7.0
Recommended: 7.4

When I tricked the upgrade script to accept 7.2, one thing that tripped me up was that composer was demanding 7.4 when certain PHP files were executed (not all). So, when these scripts were loaded from another page, the server returned 500 errors and caused a lot of things not to function properly. Examples: cron.php & AJAX.php.

**Reply by jonyo** [reply]
Yeah, I think "past me" also convinced me...  I was working on the upgrade instructions and came across part that says something like "min requirement technically 7.0 but we recommend at least 7.1 since earlier versions are no longer supported by PHP" or something like that.

So I can make the min 7.0 or maybe 7.2 - with message like mentioned - but maybe I can look into some kind of scanner or something.  See if there is a tool available that can scan the files to look for anything that only works in a higher version... That way we'll know for sure if it would actually work or not.  I dunno if something like that exists but I can look for one.

**Comment by vicos59** [comment]
`I dunno if something like that exists but I can look for one.`

Sounds like a lot of extra work. But, it's your call.

Maybe make the first v20.0.0 require 7.2 and recommend 7.4. Then post a notice that v 20.1.0 will require 7.4. Or just go with 7.4. Whatever the min was for 18.02 would probably work for 20.0.0 ?

It's your call. I have 7.4 now, so I'm safe either way.

Frankly, I spent twice as much time whining about it as it took me to research it and get a 7.4 instance running. lol

**Reply by jonyo** [reply]
> Sounds like a lot of extra work. But, it's your call.

I started down the path...  One tool reports over 2000 errors.  Skimming over them, they are all little micro bugs that have probably been in the software for years.  For instence, there is a `?a=22` version of the "extra pages".  Today extra pages uses a=26 but at some point it was using 22, or someone started coding it and did not finish, or who knows what.  Anyways, so now if you go to an extra page, then change 26 to 22 in the URL you'll see the error.

Another spot would not cause an actual error, it used uppercase name instead of lowercase.  Which PHP does not care about, but it's still good to have the class names the correct case.  So it's over 2000 of those types of things, some may be micro bugs, others things that probably don't break anything but it is good to fix.

It will be worth it to go over these over time I think.  But that is getting off a tangent to the goal, checking PHP compatibility.

I ran across another tool that will scan for syntax errors.  I think that might be the way to go.  The thing is you have to have the version of PHP you want to test, so it isn't super fast to set up.  If I have some time I might sit down and do it.

By the way, looks like it was 5.4.0 on 18.02.  Maybe a good plan is to make version 20 the "compatibility release" - people update to that version, it still mostly works on older PHP versions... Once they are successfully updated to that version, they work on getting their server up to snuff.  Then update to whatever is latest that may require 7.4.  Or maybe even 8.0 if we are able to pull of PHP 8 compatibility in the next year...

> Frankly, I spent twice as much time whining about it as it took me to research it and get a 7.4 instance running. lol

😂

**Reply by vicos59** [reply]
"compatibility release" makes good sense.

For the next jump, I would stick with 7.4. Leaping to 8.0 that soon might be a little much.

**Reply by jonyo** [reply]
My personal goal was maybe 8.0 before 7.4 is officially EOL which is November 28 this year.

Or maybe not minimum version of 8 by then, but at least "compatible with" 8 by then.

The thing is, there is a lot of really nice things that you can now do in PHP 8 that is new in that version.  Stuff I'm already doing in my "day job" since we upgraded to 8.0 a while back on all our apps.  So once we do make the minimum version 8.0 it will simplify development.  In other words, I can stop accidentally using PHP 8 stuff then wondering why the page broke 😆 .

**Reply by vicos59** [reply]
Based on your latest arguments, I would concur with making the min 8.0 for the subsequent release after the compatibility release. I know 8.0 and 8.1 are out and 8.2 is on the way. Most modern hosting services surely provide these and many control panels allow you to select the version of PHP you want to run.
